8289943: Simplify some object allocation merges

JohnTortugo opened the pull request:
Hi there, can I please get some feedback on this approach to simplify object allocation merges in order to promote Scalar Replacement of the objects involved in the merge?

The basic idea for this [approach was discussed in this thread](https://mail.openjdk.java.net/pipermail/hotspot-compiler-dev/2022-April/055189.html) and it consists of: 
1) Identify Phi nodes that merge object allocations and replace them with a new IR node called ReducedAllocationMergeNode (RAM node).
2) Scalar Replace the incoming allocations to the RAM node.
3) Scalar Replace the RAM node itself.

There are a few conditions for doing the replacement of the Phi by a RAM node though - Although I plan to work on removing them in subsequent PRs:

- The only supported users of the original Phi are AddP->Load, SafePoints/Traps, DecodeN.

These are the critical parts of the implementation and I'd appreciate it very much if you could tell me if what I implemented isn't violating any C2 IR constraints:

- The way I identify/use the memory edges that will be used to find the last stored values to the merged object fields.
- The way I check if there is an incoming Allocate node to the original Phi node.
- The way I check if there is no store to the merged objects after they are merged.

Testing:
- Windows/Linux/MAC fastdebug/release 
  - hotspot_all
  - tier1
  - Renaissance
  - dacapo
  - new IR-based tests

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [ ] Change must be properly reviewed (2 reviews required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer), 1 [Author](https://openjdk.org/bylaws#author))

### Issue
 * [JDK-8289943](https://bugs.openjdk.org/browse/JDK-8289943): Simplify some object allocation merges


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/9073/head:pull/9073` \
`$ git checkout pull/9073`

Update a local copy of the PR: \
`$ git checkout pull/9073` \
`$ git pull https://git.openjdk.org/jdk pull/9073/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 9073`

View PR using the GUI difftool: \
`$ git pr show -t 9073`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/9073.diff">https://git.openjdk.org/jdk/pull/9073.diff</a>

</details>

JohnTortugo commented:
Hi there, can someone please take a look and let me know if this is going in a reasonable direction?

@vnkozlov - is this more or less on the lines of what you were thinking?

vnkozlov reviewed (commented):
This is good starting point.  To have new "Phi" type node to collect information about merged allocation.

I need more time to dive into changes to give review.

I currently found one issue we need discuss - merge allocation of different subclasses (of the same parent class) which may have different number of fields. Current implementation assume objects are the same but I don't see the check for it during RAM node creation. May be we should have it at this initial implementation.

What about `adjust_scalar_replaceable_state()` code mark allocation as non-SR if they are merged? 

About input memory slices. Since merged allocation are SR we should have some new memory Phi created in EA `split_memory_phi()` which we can try to identify instead of adding all memory slices we find (I am talking about RAM constructor).

I see you bailed compilation to recompile in case you can't remove RAM node. I think it is fine for initial implementation.

JohnTortugo commented:
> This is good starting point. To have new "Phi" type node to collect information about merged allocation.
> I need more time to dive into changes to give review.

Thanks for taking the time to look into this!

> I currently found one issue we need discuss - merge allocation of different subclasses (of the same parent class) which may have different number of fields. Current implementation assume objects are the same but I don't see the check for it during RAM node creation. May be we should have it at this initial implementation.

Got it. I'll create some tests for this and see what happens.

> What about adjust_scalar_replaceable_state() code mark allocation as non-SR if they are merged?

I didn't get this part. Can you please clarify?

> About input memory slices. Since merged allocation are SR we should have some new memory Phi created in EA split_memory_phi() which we can try to identify instead of adding all memory slices we find (I am talking about RAM constructor).

I'll take a look into that. Thanks for the suggestion.

> I see you bailed compilation to recompile in case you can't remove RAM node. I think it is fine for initial implementation.

Great.

vnkozlov commented:
> > What about adjust_scalar_replaceable_state() code mark allocation as non-SR if they are merged?
> 
> I didn't get this part. Can you please clarify?

My bad. After looking more on changes I noticed that you exit `compute_escape()` before `adjust_scalar_replaceable_state()` is called. So my comment is null.

JohnTortugo commented:
> After looking more on changes I noticed that you exit compute_escape() before adjust_scalar_replaceable_state() is called. 

No worries. I'm currently working to make `reduce_allocation_merges` be executed as part of compute_escape so that we can take benefit from iterative EA executions.

navyxliu commented on src/hotspot/share/opto/callnode.cpp:
```diff
@@ -1614,6 +1618,233 @@ Node *AllocateArrayNode::make_ideal_length(const TypeOopPtr* oop_type, PhaseTran
+  return true;
+}
+
+bool ReducedAllocationMergeNode::register_use(Node* n) {
```
I think register_use/addp() don't fail in runtime.  it will simplify your code.

navyxliu commented on src/hotspot/share/opto/escape.cpp:
```diff
@@ -387,6 +401,256 @@ bool ConnectionGraph::compute_escape() {
+
+  _igvn->_worklist.push(ram);
+
+//  if (n->_idx == 257 && ram->_idx == 1239) {
```
some mysterious code? :)

JohnTortugo replied:
Oops (no pun intended).

navyxliu commented on src/hotspot/share/opto/callnode.hpp:
```diff
@@ -1008,6 +1011,63 @@ class AllocateArrayNode : public AllocateNode {
+// merging the value produced in the different inputs to the RAM node.
+// In some cases a reference to the whole merged object is needed and
+// we handle that by creating an SafePointScalarObjectNode.
+class ReducedAllocationMergeNode : public TypeNode {
```
Is that possible to make it just subclass of PhiNode?  if so, you wouldn't need to touch cfgnode.[hpp/cpp]

JohnTortugo replied:
@navyxliu - Thanks for commenting. I considered that option and it seems intuitive that RAM should be a subclass of Phi. However, in terms of implementation, the similarity isn't that big. For now, I decided to not make it a subclass of Phi, to prevent code that only checks `is_Phi` from inadvertently changing the edges of RAM. Of course, we can revisit this decision.

navyxliu commented on src/hotspot/share/opto/c2_globals.hpp:
```diff
@@ -471,6 +471,9 @@
   develop(bool, TracePostallocExpand, false, "Trace expanding nodes after"  \
           " register allocation.")                                          \
                                                                             \
+  product(bool, ReduceAllocationMerges, false,                              \
```
we need enable it anyway.

JohnTortugo commented:
Thanks for taking a look @navyxliu . I'm working on an improvement (and fixing a bug) and I'll include your suggestions in the next push.

JohnTortugo commented:
Hi @vnkozlov , @navyxliu - I think I addressed all your comments so far. Can I ask you to please take a look again? I made several modifications for fixing bugs and improving the coverage of the optimization. In this new version the simplification of allocations is performed in every EA iteration, the memory edges in RAM are updated after split_unique_types is computed and only allocations merged inputs of the same `_klass` are simplified.

I tested this in several benchmarks and didn't find any regression. Nevertheless, I'm still going to run more tests with different JVM configurations.

Can I ask you (@TobiHartmann , @rwestrel , @iwanowww) to also take a look?

JohnTortugo commented:
>I currently found one issue we need discuss - merge allocation of different subclasses (of the same parent class) which may have different number of fields. Current implementation assume objects are the same but I don't see the check for it during RAM node creation. May be we should have it at this initial implementation.

Hi again @vnkozlov - The current implementation requires the Klass of all inputs to the Phi to be the same. I'm trying to lift that restriction as it would increase the number of cases where the merge can be removed. I've been trying to create an example where I'd need to enforce the allocations to be of the same type but I couldn't figure one out yet. Do you have an example where that must be enforced? Note that I only need fields accessed through the merge Phi node, which means only fields present in the superclass AFAIU.

vnkozlov commented:
Hi @JohnTortugo 
Yes, I think you can lift restriction for all inputs have the same class. I looked more and you correctly set RAM's type to original Phi's type - it should be correct type (base class, which could be also one of inputs). You correctly register fields offsets based on this (Phi) class during RAM make.  C2 type system and bytecode should guarantee that Phi's type is correct (meet of all inputs). It could be more narrowed during CCP phase but type you got serves correctly for this implementation. May be assert that it is not interface - or skip such cases if you find them.

You can add `assert(input_t->is_instptr()->instance_klass()->is_subtype_of(klass))` instead of klass equals check.

What confused me is code in `PhaseMacroExpand::scalar_replacement()` where you process Allocation from one of inputs which may have more fields than RAM registered. But you correctly have `!ram->needs_field(offset)` check which I missed before. That why I commented.

vnkozlov commented:
Allocations in `testPollutedPolymorphic()` are removed because both classes have the same `Shape` class which have all fields. Would be interesting if `l` field is declared only in both subclasses.

JohnTortugo commented:
@vnkozlov - Thank you for clarifying that. I've been playing with lifting the restriction and I actually found a corner case:

```
public static Class test(boolean c1, boolean c2, boolean c3, int x, int y, int w, int z) {
    Animal s = new Dog(x, y, z);

    if (c1) {
        s = new Cat("Fisker");
    }

    Unloaded u = new Unloaded(); // assumes this is converted to a uncommon_trap(unloaded, reinterpret)

    return s.getClass();
}
```

It seems that when merging allocations of different subtypes I'll need to add a special `Phi` node merging the `Klass` of the input allocations. If I don't do that, the method above will return Animal.class instead of `Dog.class` or `Cat.class`. I'm wondering if I'll actually have to do the same for the Header/Mark word of the input allocations.

vnkozlov commented:
@JohnTortugo Yes, you would need to construct Phi when you replace RAM with `SafePointScalarObjectNode`. Hmm, may be you would need to construct Phi in other cases too (getClass intrinsic).

Add cases when class is loaded from argument for allocation: `Unsafe.allocateInstance()` and `Object.clone()` to test class Load instead of simple constant on some paths of such Phi.

Why you need Phi for mark word? For identity check (hashCode/identityHashCode intrinsics)?

JohnTortugo commented:
> @JohnTortugo Yes, you would need to construct Phi when you replace RAM with SafePointScalarObjectNode. Hmm, may be you would need to construct Phi in other cases too (getClass intrinsic).

Yes. I'll take a look into `getClass` intrinsic. I thought that just adding input to `SafePointScalarObjectNode`+Safepoint with a Phi of the input allocations Klass fields would be enough for the code to correctly access fields/methods of the input objects. However, it looks like the Klass of the object rematerialized from SafePointScalarObjectNode is [the type of `SafePointScalarObjectNode`](https://github.com/openjdk/jdk/blob/master/src/hotspot/share/opto/output.cpp#L859), not the Klass field input edge set in the Safepoint.

> Add cases when class is loaded from argument for allocation: Unsafe.allocateInstance() and Object.clone() to test class Load instead of simple constant on some paths of such Phi.

I'll do that. Thanks for the tip.

> Why you need Phi for mark word? For identity check (hashCode/identityHashCode intrinsics)?

Yes, I was wondering about a case where I'd need any of the information present in the Mark word. The code in the PR is able to solve merges where only some of the merged allocations are removed. I believe I'll need, at least, to preserve the Mark word of the allocations not removed.

JohnTortugo commented:
Hi again @vnkozlov, I ended up giving up on adding support for merging allocations of different subclasses. I think the current version of the patch already covers some reasonable ground and it seems that adding support for merging allocations of subclasses will require more widespread changes. 

Also, I wasn't able to find a case where the patch fail due to an incorrect/missing Mark word. FYI we have been testing the current patch with all JTREG tests, TCK tests, Renaissance, DaCapo, and some customer services. I currently don't see any failure.

I also added a few JMH tests to show up what kind of improvements we can expect with the proposed changes:

```
Benchmark                                                         Mode  Cnt         Score        Error  Units

AllocationsMerge.IfElseInLoop                                     avgt    5  93,839,709.289 ± 16846405.548  ns/op
AllocationsMerge.MergeAndIterative                                avgt    5     160,988.682 ±    57389.599  ns/op
AllocationsMerge.NestedObjectsObject                              avgt    5     311,523.642 ±    59622.703  ns/op
AllocationsMerge.SimpleMerge                                      avgt    5     142,398.057 ±    57010.144  ns/op
AllocationsMerge.TrapAfterMerge                                   avgt    5     333,061.738 ±    25645.896  ns/op

AllocationsMerge.WithAllocationsMergeEnabled.IfElseInLoop         avgt    5  11,964,798.596 ±  1661843.496  ns/op
AllocationsMerge.WithAllocationsMergeEnabled.MergeAndIterative    avgt    5      33,118.428 ±     2995.224  ns/op
AllocationsMerge.WithAllocationsMergeEnabled.NestedObjectsObject  avgt    5     155,808.370 ±    30379.710  ns/op
AllocationsMerge.WithAllocationsMergeEnabled.SimpleMerge          avgt    5      31,266.188 ±    13476.634  ns/op
AllocationsMerge.WithAllocationsMergeEnabled.TrapAfterMerge       avgt    5     208,694.038 ±    18667.023  ns/op
```

Please let me know if there is something else that I can do to make it easier to review the proposed changes.

vnkozlov commented on src/hotspot/share/opto/escape.cpp:
```diff
@@ -387,6 +398,318 @@ bool ConnectionGraph::compute_escape() {
+          NOT_PRODUCT(if (TraceReduceAllocationMerges) tty->print_cr("Will NOT try to reduce Phi %d. DecodeN use is not a Load. %d %s", phi->_idx, use_use->_idx, use_use->Name());)
+          return false;
+        }
+
```
You are missing check for constant offset as you did in previous code. May be move this code into small method.
Also this whole code which checks users could be separate method.

vnkozlov commented on src/hotspot/share/opto/escape.cpp:
```diff
@@ -387,6 +398,318 @@ bool ConnectionGraph::compute_escape() {
+    }
+  }
+
+  if (mixed_klasses && has_call_as_user) {
```
Add comment that only the same Klass is allowed if merged allocations are referenced by SafePoint or Uncommon
trap.

vnkozlov replied:
You also need to bail out `has_nonnull_nonsr_input` case here. You can have not Allocation in such case (for example, argument) which you can't reallocate during deoptimization and you don't know its actual Klass.

vnkozlov commented on src/hotspot/share/opto/callnode.cpp:
```diff
@@ -1685,6 +1689,277 @@ Node *AllocateArrayNode::make_ideal_length(const TypeOopPtr* oop_type, PhaseTran
+  else if (n->is_DecodeN()) {
+    for (DUIterator_Fast imax, i = n->fast_outs(imax); i < imax; i++) {
+      Node* addp = n->fast_out(i);
+      assert(addp->is_AddP(), "DecodeN user is not an AddP");
```
You don't need this assert since `as_AddP()` has it.

vnkozlov commented on src/hotspot/share/opto/escape.cpp:
```diff
@@ -387,6 +398,318 @@ bool ConnectionGraph::compute_escape() {
+      mixed_klasses = true;
+    }
+
+    klass = input_t->is_instptr()->instance_klass();
```
Why you are updating `Klass` value here?

JohnTortugo replied:
My mistake. No need for that.

vnkozlov reviewed (commented):
Please, merge the latest JDK sources after you address my comments. I can submit testing after that.
And you will need second review.

JohnTortugo commented:
Thank you for taking a look @vnkozlov . I'll address your comments asap.

JohnTortugo commented:
Hi @vnkozlov - I pushed some changes that I think address the comments that you made. I also had to add another constraint to reducing the allocations: for now, I won't be reducing merges of SR and NSR allocations. I'll definitely work on that in a subsequent patch.

Thank you again for reviewing and I'd appreciate it very much if you could check if the patch doesn't break any of your tests.

vnkozlov commented on src/hotspot/share/opto/c2_globals.hpp:
```diff
@@ -475,6 +475,12 @@
   develop(bool, TracePostallocExpand, false, "Trace expanding nodes after"  \
           " register allocation.")                                          \
                                                                             \
+  product(bool, ReduceAllocationMerges, false,                              \
+          "Try to simplify allocation merges before Scalar Replacement")    \
```
I don't see flag's value change to `true` as suggested.

vnkozlov commented on src/hotspot/share/opto/callnode.cpp:
```diff
@@ -1665,6 +1669,276 @@ Node *AllocateArrayNode::make_ideal_length(const TypeOopPtr* oop_type, PhaseTran
+                    memory->in(b_idx) :
+                    memory);
+    }
+  }
+  else {
```
Style - should be one line: ` } else {`.
I noticed that here and in other places in new code you put `else` and `else if (` on separate line from `}`. We use one line in such cases.

vnkozlov commented on src/hotspot/share/opto/escape.hpp:
```diff
@@ -575,6 +582,32 @@ class ConnectionGraph: public ResourceObj {
+  //  - All inputs should be NoEscape
+  //  - The only uses of the Phi should be:
+  //    - AddP->Load
+  //    - SafePointNode or uncommon traps
+  //    - DecodeN->AddP->Loads
+  bool can_reduce_this_phi(const Node* n) const;
```
Move detailed description comment to .cpp file.

vnkozlov commented on src/hotspot/share/opto/escape.hpp:
```diff
@@ -626,6 +667,24 @@ class ConnectionGraph: public ResourceObj {
+  //
+  // JavaObject(3) NoEscape(NoEscape) Edges: [ 91F ----       ] Uses: [ 40 45 ---   ]   28  Allocate ...
+  // JavaObject(4) NoEscape(NoEscape) Edges: [ 172F ----      ] Uses: [ 119 124 --- ]  107  Allocate ...
+  // LocalVar(12)  NoEscape(NoEscape) Edges: [ 119 107P       ] Uses: [ ---         ]  124  CheckCastPP ...
+  // LocalVar(11)  NoEscape(NoEscape) Edges: [ 40 28P         ] Uses: [ ---         ]   45  CheckCastPP ...
+  void remove_phi_node(PointsToNode* ptn);
```
Move detailed description comment to .cpp file.

vnkozlov commented on src/hotspot/share/opto/callnode.cpp:
```diff
@@ -1665,6 +1669,276 @@ Node *AllocateArrayNode::make_ideal_length(const TypeOopPtr* oop_type, PhaseTran
+    field_type = Type::get_const_basic_type(basic_elem_type);
+  }
+
+  Node* load = LoadNode::make(*igvn, ctrl, mem, addp, adr_type, field_type, basic_elem_type, MemNode::unordered,
+                                LoadNode::DependsOnlyOnTest, false, false, false, false, (uint8_t)0U, false);
```
Add comment explaining why you want to keep `LoadN` node without `DecodeN`.

vnkozlov reviewed (commented):
Thank you for addressing my comments. I have mostly code style related comments left (and default flag's value).
I will submit testing and we will see how it goes.

vnkozlov commented:
@JohnTortugo  Good news is that I ran from tier1 to tier4 and with `-Xcomp` with `ReduceAllocationMerges` enabled and I did not get new failures except in your new test `compiler/c2/irTests/scalarReplacement/AllocationMergesTests.java` when it run with `-ea -esa -XX:CompileThreshold=100 -XX:-TieredCompilation` flags.

I put failed test's output to RFE.

vnkozlov commented:
/reviewers 2

vnkozlov commented:
We need others to look on this too at this stage.

vnkozlov commented:
And I will submit performance runs.

JohnTortugo commented:
Thanks a lot for the comments @vnkozlov . I'll push new changes addressing your suggestions later today.

merykitty commented on test/hotspot/jtreg/compiler/c2/irTests/scalarReplacement/AllocationMergesTests.java:
```diff
@@ -0,0 +1,939 @@
+        if (cond)
+            p = new Picture(y, y, x);
+
+        return p.position;
```
Hi, you may want to add a test case in which this field does not escape. Also, since this is a really large test, reducing the number of warmup iterations will reduce the runtime of the test. Thanks.

vnkozlov commented:
> And I will submit performance runs.

Performance results are mixed bag. Saw some improvements on x64 in Renaissance but they gone after rerun.
I see some significant regression in Dacapo on Linux Aarch64 (Ampere CPU) but don't see it at all on MacOS M1.

JohnTortugo commented:
Hi @vnkozlov  & @merykitty - thank you for the reviews. I'm working on those. I had to take some time off work due to personal issues but I believe I can address your comments during the weekend or earlier next week.

@vnkozlov - Thank you so much for running the perf. tests! Can you please give me more details about the experiment? What metric did you analyze, which version of Dacapo, and was it any specific benchmark that regressed or all of them? Which JVM parameters?

Regards,
Cesar

vnkozlov commented:
DaCapo 9.12 I think.
`DaCapo-xalan-large` showed -36% and -27% in 2 runs. Flags: `--size large --iterations 20 xalan`
`DaCapo-lusearch-large` showed -19% and -114% in 2 runs. Flags: `--size large --iterations 20 lusearch-fix`

@ericcaspole said that Dacapo results are not stable.

On MacOS M1 there is no difference in results (insignificant <0.5%).

JohnTortugo commented:
Hi @vnkozlov, @merykitty  - I pushed some changes to address your suggestions.

@vnkozlov - Thank you for the additional info about DaCapo. Looks like you're doing a comparison based on CPU or Wall Time, is that so? I did some further investigation on my end and I didn't see much change in Wall Time or Allocation Rate in Xalan or LUSearch. The DaCapo benchmarks don't allocate a lot of memory and in my experiments so far I don't see the current patch having a big impact on the performance of those benchmarks. I'm wondering if the effect that you're experiencing is due to some unrelated effect? _Still, I admit that I only did performance tests on Linux since I expected my changes to be independent of the target architecture and/or OS. I'm going to run additional tests on different targets and investigate any regression that I may find._

In a previous batch of perf. experiments I saw noticeable gains in some Renaissance benchmarks (experiments run on Linux). Something around a 10% reduction in allocation rate in a few benchmarks and in others I didn't see any statistical difference. I'm going to run some more perf. experiments and in more platforms/OS and get back to you. If there is any particular benchmark or argument that you think it's important for me to use, please let me know.

I'm considering making the `ReduceAllocationMerges` an experimental option. I think that makes it more clear that we have a stable patch but we are still actively working to make it more widely applicable and thus increasing its positive impact on performance. What do you think?

vnkozlov commented:
The only is issue is that experimental flags usually have `false` default value. Alternatively you can make it diagnostic which we use for C2 optimizations flags.

JohnTortugo commented:
My idea was to have it as `false` by default. TBH, I thought that was kind of the "policy" for adding new optimizations.

vnkozlov commented:
DaCapo and Renaissance are good for testing it I think. That is where I see variations. May be we can try to use `-XX:-TieredCompilation` to see if using only C2 have effect.

It seems we don't have a lot of cases where this optimization helps. May for future work based on these benchmarks (and others) we can collect cases when this optimization does not work (or even bailout compilation).

 BTW, were you able to remove all allocations in your test `run_IfElseInLoop()`?
What about  test case in https://bugs.openjdk.org/browse/JDK-6853701

vnkozlov commented:
> My idea was to have it as `false` by default. TBH, I thought that was kind of the "policy" for adding new optimizations.
Yes, experimental flag is good for now if you DO plan to switch it ON later (and convert flag into other type).
The trouble with `false` experimental flags is that code become dead and "rot" if nobody care about it.

vnkozlov commented:
@JohnTortugo I just noticed that both new tests missing `{}` in a lot of conditional cases. Please, fix it. Code style.

JohnTortugo commented:
> Yes, experimental flag is good for now if you DO plan to switch it ON later (and convert flag into other type).

Yes, our plan is to keep working on this until we can remove all the constraints in `can_reduce_this_phi`. We have already discussed ideas of how to remove those constraints and we'll be tackling that (and discussing it with you & the community) after we get this initial implementation upstreamed.

> The trouble with false experimental flags is that code become dead and "rot" if nobody care about it.

Totally understand your concern. I'm fine with setting the flag to true and making it "Diagnostic" - I just don't see it as a "Diagnostic" patch, but if it's fine for you then it's fine for me.

JohnTortugo commented on test/hotspot/jtreg/compiler/c2/irTests/scalarReplacement/AllocationMergesTests.java:
```diff
@@ -0,0 +1,982 @@
+    }
+
+    @Test
+    @IR(failOn = { IRNode.ALLOC })
```
@vnkozlov - Both allocs will be removed.

JohnTortugo commented:
> May for future work based on these benchmarks (and others) we can collect cases when this optimization does not work (or even bailout compilation).

The `TraceReducedAllocationMerges` option prints information about this. I actually have a spreadsheet where I list the cause and frequency of each case where the optimization can not be applied.

> BTW, were you able to remove all allocations in your test run_IfElseInLoop()?

Yes, in that case both allocations are removed. I just confirmed it with a test locally. Also, there is an IR-based test for that case.

> What about test case in https://bugs.openjdk.org/browse/JDK-6853701

The current patch bails out in that test because there is a Phi (or CmpP) consuming the merge Phi. Actually, that code example is one of the tests that I run "internally". There is already work going on to improve the current patch to make it able to handle CmpP with NULL.

vnkozlov commented:
Thank you for answering my questions and doing all research.

Let do this: make flag "experimental" and `true` by default. This will allow to test it for some time after changes are integrated. If everything looks good we keep it that way otherwise we can switch it off before JDK 20 is shipped (or switch it off regardless before shipping).

I will start new round of testing (more tiers) and I will ask others to review changes.
I think we should go with current state of changes if it satisfied other reviewers.

iwanowww reviewed (commented):
Very nice work, Cesar. Most notably, I'm happy to see the test with so many non-trivial cases enumerated.

Speaking of the proposed patch itself, I'm not done yet reviewing it.

As of now, I don't fully grasp what's the purpose and motivation to introduce `ReducedAllocationMerge`. I would be grateful for additional information about how you ended up with the current design. (I went through the email thread, but it didn't help me.)

In particular, I still don't understand how it interacts with existing scalar replacement logic when it comes to unique (per-allocation) memory slices.

Also, on bailouts when the new analysis fails: I instrumented all possible failure modes with asserts and all the test failures I saw were in 2 places:
```
src/hotspot/share/opto/callnode.cpp:1912
    } else if (input->bottom_type()->base() == Type::Memory) {
      // Somehow the base was eliminated and we still have a memory reference left
      assert(false, "");
      return NULL;
    }

src/hotspot/share/opto/macro.cpp:2612
        // In some cases the region controlling the RAM might go away due to some simplification
        // of the IR graph. For now, we'll just bail out if this happens.
        if (n->in(0) == NULL || !n->in(0)->is_Region()) {
          assert(false, "");
          C->record_failure(C2Compiler::retry_no_reduce_allocation_merges());
          return;
        }
```
How hard would it be to extend the test with cases which demonstrate existing limitations?

iwanowww commented:
Additional high-level comments.

(1) Overall, I don't see (yet) enough motivation to justify the addition of `ReducedAllocationMerge`. I'd prefer to see the new node go away.
 
`ReducedAllocationMerge` nodes are short-lived. As I can infer from the code, they are used solely for the purpose of caching field information (in addition to marking that the original phi satisfied some requirements). Have you considered using existing `ConnectionGraph` instance for bookkeeping purposes? It's available during IGVN and macro expansion. 

Also, I believe you face some ideal graph inconsistencies because you capture information too early (before `split_unique_types` and following IGVN pass; and previous allocation eliminations during `eliminate_macro_nodes()` may contribute to that).

(2) Following up on my earlier question about interactions with `split_unique_types()`, I'm worried that you remove corresponding `LocalVar`s from the ConnectionGraph and introduce with unique memory slices. I'd feel much more confident in the correctness if you split slices for unions of interacting allocations instead.

iwanowww commented on src/hotspot/share/opto/escape.cpp:
```diff
@@ -387,6 +398,330 @@ bool ConnectionGraph::compute_escape() {
 
+int ConnectionGraph::reduce_allocation_merges() {
+  Unique_Node_List ideal_nodes;
+  ideal_nodes.map(_compile->live_nodes(), NULL);
```
You do a full pass over the whole graph to enumerate `Phi`s. Don't you achieve the same when enumerating `LocalVar`s from the CG?

iwanowww commented on src/hotspot/share/opto/escape.cpp:
```diff
@@ -3634,47 +4050,75 @@ void ConnectionGraph::split_unique_types(GrowableArray<Node *>  &alloc_worklist,
-    assert(old_cnt == old_mem->outcnt(), "old mem could be lost");
+
+  // Update the memory inputs of ReducedAllocationMerge nodes
+  Unique_Node_List ram_nodes;
```
Same here: pass over the whole graph to enumerate `ReducedAllocationMerge` nodes. For such cases, we keep node lists on Compile (see `GrowableArray<Node*>` declared there).

iwanowww reviewed (commented):
Minor comments. No need to act until we agree on high-level aspects of the patch.

iwanowww commented on src/hotspot/share/opto/escape.cpp:
```diff
@@ -387,6 +398,330 @@ bool ConnectionGraph::compute_escape() {
+
+          if (child->is_Store()) {
+            assert(child->in(0) != NULL || m->in(0) != NULL, "No control for store or AddP.");
+            if (_igvn->is_dominator(merge_phi_region, child->in(0) != NULL ? child->in(0) : m->in(0))) {
```
`PhaseIterGVN::is_dominator()` is a conservative approximation and can result in false negatives.

JohnTortugo commented:
Hi Vladimir, first of all, thank you for reviewing the proposed patch. Sorry for
the delay answering your questions, I'm right now on paternity leave from work
and my time in front of a computer is being quite limited.

> As of now, I don't fully grasp what's the purpose and motivation to introduce
> ReducedAllocationMerge. I would be grateful for additional information about how
> you ended up with the current design. (I went through the email thread, but it
> didn't help me.)

My first implementation to deal with the problem was just replacing Phi's
merging object allocations by sets of Phi's merging field's loads from each
different base. That works fine in some cases. However, one challenge that I
faced in this approach, IIRC, was that the new Phi nodes (merging loads of
fields) were being factored away (eliminated) as part of IGVN optimizations and
consequently the graph ended up again with Phi's merging object allocations. I
considered inserting Opaque nodes in some branches of the graph to prevent
unadverted optimizations but at the end I found the approach below more
"robust".

The proposed idea of using a new type of node to represent the merges have
these Pros, IMO.  

- By replacing Phi's merging allocations by a new type of node (i.e., RAM) the
existing code (split_unique_types) will be able to assign `instance_id`s to the
scalar replaceable inputs of the merging Phi. That means that those inputs will
be scalar replaced using existing code in C2. Additional code will be necessary
only to scalar replace the new type of node.

- As a side effect of the point described above, we'll also be able use existing
code in C2 to find last stored values to fields (i.e., find_inst_mem,
value_from_mem, etc).

- Existing optimizations will not interfere with the scalar replacement (as they
did in the previous approach outlined) because they don't know how to handle the
new type of node. In essence, the new type of node will be Opaque to existing
optimizations.

- We can safely, AFAIU, run igvn.optimize() after replacing the allocation merge
Phis with the new type of node.

- For last, an additional benefit of using a new type of node to store
information, _instead_ of storing it in the ConnectionGraph, for instance, is
that by using graph edges to capture the required "information" that the node
need, the value is never in an "outdated" state. For instance, the current patch
use input slots of the RAM node to store reference to required memory edges.
Anytime a transformation happens in the graph and affect that memory edge the
RAM node will be "automatically" updated using existing code in C2. If, instead,
the memory edge was just stored as part of internal data of some class, then
we'd need to handle those updates manually, AFAIU.
 
Some arguably Cons of the current approach:

- Seems complex at first glance.

- New (non functional) node in the IR. The node is quite similar to a
PhiNode in the sense that it's there just to represent a state (or some set of
information).

- The new node is a Macro node. Failure to remove it from the IR graph can cause
compilation failure.


> In particular, I still don't understand how it interacts with existing scalar
> replacement logic when it comes to unique (per-allocation) memory slices.

The RAM node itself doesn't interfere with alias indexes / memory slices
creation at all. RAM nodes are created before "adjust_scalar_replaceable_state"
is executed and because of that the allocations participating in merges _may_ be
marked as ScalarReplaceable. Later, "split_unique_types" will run and be able to
assign instance_id's to allocations participating in the merge because there is
"virtually" no merge anymore at this point.

During the Macro node elimination phase the allocation nodes will be visited and
potentially scalar replaced _before_ any RAM node is visited. When an allocation
node being scalar replaced is consumed by a RAM node some information are
"registered" in the RAM node. Those information will later be used when the time
come to scalar replace the RAM node itself.

I hope I have answered your question. If not please let me know and I'll be
happy to give more details.


> How hard would it be to extend the test with cases which demonstrate existing
> limitations?

I'll try and create some test cases that trigger those limitations.


> Also, I believe you face some ideal graph inconsistencies because you capture
> information too early (before split_unique_types and following IGVN pass; and
> previous allocation eliminations during eliminate_macro_nodes() may contribute
> to that).

Can you please elaborate on that?


> Following up on my earlier question about interactions with
> split_unique_types(), I'm worried that you remove corresponding LocalVars from
> the ConnectionGraph and introduce with unique memory slices. I'd feel much more
> confident in the correctness if you split slices for unions of interacting
> allocations instead.

Can you please explain a bit more about this idea? Are you proposing to split the
phi into slices instead of removing it?

iwanowww commented:
Thanks for the clarifications, Cesar.

The concept of RAM node still looks like a hack to me. And I'd like the patch to better fit the overall design of EA rather than trying to workaround different peculiarities of current implementation.

I'll try to elaborate why I see RAM node redundant.

As of now, it serves dual purpose. It (1) marks a merge point as safe to be untangled during SR; and (2) caches information about field values.

I believe you can solve it in a cleaner manner without introducing placeholder
nodes and connection graph adjustments. IMO it's all about keeping escape 
status and properly handling "safe" merges in split_unique_types.

One possible way to handle merge points is:
  * Handle merge points in adjust_scalar_replaceable_state and refrain from marking relevant bases as NSR when possible.
  * After adjust_scalar_replaceable_state is over, every merge point should have all its inputs as either NSR or SR. 
  * split_unique_types incrementally builds value phis to eventually replace the base phi at merge point while processing SR allocations one by one.
  * After split_unique_types is done, there are no merge points anymore, each allocation has a dedicated memory graph and allocation elimination can proceed as before.

Do you see any problems with such an approach?

One thing still confuses me though: the patch mentions that RAMs can merge both eliminated and not-yet-eliminated allocations. What's the intended use case? I believe it's still required to have all merged allocations to be eventually eliminated. Do you try to handle the case during allocation elimination when part of the inputs are already eliminated and the rest is pending their turn?

iwanowww commented:
>> Also, I believe you face some ideal graph inconsistencies because you capture
information too early (before split_unique_types and following IGVN pass; and
previous allocation eliminations during eliminate_macro_nodes() may contribute
to that).

> Can you please elaborate on that?

There are places in the patch where you check for unintended graph modifications, like in `PhaseMacroExpand::eliminate_macro_nodes()`:
```
        // In some cases the region controlling the RAM might go away due to some simplification
        // of the IR graph. For now, we'll just bail out if this happens.
        if (n->in(0) == NULL || !n->in(0)->is_Region()) {
          C->record_failure(C2Compiler::retry_no_reduce_allocation_merges());
          return;
        }
```

I consider that an implementation peculiarity contributed by RAMs rather than an inherent complication coming from the transformation being performed.

vnkozlov commented:
> 
> As of now, it serves dual purpose. It (1) marks a merge point as safe to be untangled during SR; and (2) caches information about field values.

An other important purpose of RAM is to have information at SafePoints after merge point for reallocation during deoptimization. You need Klass information. I don't think having only Phis for values is enough.

> 
> I believe you can solve it in a cleaner manner without introducing placeholder nodes and connection graph adjustments. IMO it's all about keeping escape status and properly handling "safe" merges in `split_unique_types`.
> 
> One possible way to handle merge points is:
> 
> * Handle merge points in `adjust_scalar_replaceable_state` and refrain from marking relevant bases as NSR when possible.
> * After `adjust_scalar_replaceable_state` is over, every merge point should have all its inputs as either NSR or SR.
> * `split_unique_types` incrementally builds value phis to eventually replace the base phi at merge point while processing SR allocations one by one.
> * After `split_unique_types` is done, there are no merge points anymore, each allocation has a dedicated memory graph and allocation elimination can proceed as before.

I am not sure how this could be possible. Currently EA rely on IGVN to propagate fields values based on unique memory slice. What you do with memory Load or Store nodes after merge point? Which memory slice you sleet for them? 

> 
> Do you see any problems with such an approach?
> 
> One thing still confuses me though: the patch mentions that RAMs can merge both eliminated and not-yet-eliminated allocations. What's the intended use case? I believe it's still required to have all merged allocations to be eventually eliminated. Do you try to handle the case during allocation elimination when part of the inputs are already eliminated and the rest is pending their turn?

There is check for it in `ConnectionGraph::can_reduce_this_phi()`. The only supported cases is when no deoptimization point (SFP or UNCT) after merge point. It allow eliminate SR allocations even if they merge with NSR allocations. This was idea.

iwanowww commented:
> An other important purpose of RAM is to have information at SafePoints after merge point for reallocation during deoptimization. You need Klass information. I don't think having only Phis for values is enough.

Klass information is available either from Allocation node in `split_unique_types` or ConnectionGraph instance the Phi is part of. 


> I am not sure how this could be possible. Currently EA rely on IGVN to propagate fields values based on unique memory slice. What you do with memory Load or Store nodes after merge point? Which memory slice you will use for them?

My understanding of how proposed approach is expected to work: merge points have to be simple enough to still allow splitting unique types for individual allocations. 

For example, `eliminate_ram_addp_use()` replaces `Load (AddP (Phi base1 ... basen) off) mem` with `Phi (val1 ... valn)` and `eliminate_reduced_allocation_merge()` performs similar transformation for `SafePoint`s. 

Alternatively, corresponding `Phi`s can be build incrementally while processing each individual `base` by `split_unique_types`. Or, just by splitting `Load`s through `Phi`: 
```
Load (AddP (Phi base_1 ... base_n) off) mem 
== split-through-phi ==>
Phi ((Load (AddP base_1 off) mem) ... (Load (AddP base_n off) mem)) 
== split_unique_types ==>
Phi ((Load (AddP base_1 off) mem_1) ... (Load (AddP base_n off) mem_n)) 
== IGVN ==>
Phi (val_1 ... val_n)
``` 

> There is check for it in ConnectionGraph::can_reduce_this_phi(). The only supported cases is when no deoptimization point (SFP or UNCT) after merge point. It allow eliminate SR allocations even if they merge with NSR allocations. This was idea.

That's nice! Now I see `has_call_as_user`-related code. 

It means that only `Load (AddP (Phi base_1 ... base_n)  off) mem` shapes are allowed now.
I believe the aforementioned split-through-phi transformation should handle it well: 
```
Load (AddP (Phi base_1 ... base_n) off) mem 
== split-through-phi ==>
Phi ((Load (AddP base_1 off) mem) ... (Load (AddP base_n off) mem)) 
== split_unique_types ==>
Phi (... (Load (AddP base_SR_i off) mem_i) ... (Load (AddP base_NSR_n off) mem) ...) 
== IGVN ==>
Phi (... val_i ... (Load (AddP base_NSR_n off) mem) ... )
```

JohnTortugo commented:
Hi @iwanowww - Thank you for clarifying things! 

After much thought and some testing, I think I can make the RAM node go away and achieve the results I want. Below are some additional comments & the overall approach that I'm going to switch to.

- `LoadNode::split_through_phi` requires `is_known_instance_field` and therefore can't be run before `split_unique_types` without changes.

-  I think splitting the merge Phi as part of `adjust_scalar_replaceable_state` might be the best place to create the value Phi for the fields. However, if the merge Phi is used by a SafePoint/UncommonTrap then it's necessary to also create a `SafePointScalarObjectNode` (SSON). As you know, there is already logic to create SSON as part of `PhaseMacroExpand::scalar_replacement`. So, we have to decide if it's best to re-use the code in `PhaseMacroExpand::scalar_replacement` in `adjust_scalar_replaceable_state` or if we want to add new code to create SSON for merge Phis in `PhaseMacroExpand::scalar_replacement`. Here are the Pros & Cons of the two options I mentioned above:
   - Create SSON in `adjust_scalar_replaceable_state`:
            Pros: all logic to split phi is centered in one place.
            Cons: the logic to create SSON is used outside `scalar_replacement` routine.

    - Create SafePointScalarReplacedNode in `scalar_replacement` routine.
            Pros: Scalar replacement of merge Phi is contained in scalar_replacement routine.
            Cons: the logic to split merge phi is spread throughout escape analysis and scalar replacement.

Does it look like a better approach? TIA!

iwanowww commented:
> Does it look like a better approach?

It definitely does! Thanks a lot for thinking it through.

> LoadNode::split_through_phi requires is_known_instance_field and therefore can't be run before split_unique_types without changes.

IMO it should be fine to customize `LoadNode::split_through_phi` for our purposes. 

> So, we have to decide if it's best to re-use the code in PhaseMacroExpand::scalar_replacement in adjust_scalar_replaceable_state or if we want to add new code to create SSON for merge Phis in PhaseMacroExpand::scalar_replacement. 

From design perspective, it's cleaner to modify the IR during `PhaseMacroExpand::scalar_replacement()`, since `ConnectionGraph::adjust_scalar_replaceable_state()` operates solely on `ConnectionGraph` instance being built.

JohnTortugo commented:
Hi all! I'm back from leave and I'm resuming the work on this project. I'll keep you updated. Thanks again for all the comments and feel free to add more in the meantime.
